api/server/controllers/ModelController.js

```diff
@@ -18,16 +18,10 @@ const getModelsConfig = async (req) => {
  * @returns {Promise<TModelsConfig>} The models config.
  */
 async function loadModels(req) {
-  const cache = getLogStores(CacheKeys.CONFIG_STORE);
-  const cachedModelsConfig = await cache.get(CacheKeys.MODELS_CONFIG);
-  if (cachedModelsConfig) {
-    return cachedModelsConfig;
-  }
-  const defaultModelsConfig = await loadDefaultModels(req);
-  const customModelsConfig = await loadConfigModels(req);
-
-  const modelConfig = { ...defaultModelsConfig, ...customModelsConfig };
+  const modelConfig = { ...(await loadDefaultModels(req)), ...(await loadConfigModels(req)) };
```

Comment:
with this change, the cache is never used for models. this will lead to slow page loads every time.

Reply:
Without this it will not work. Now what?

The cache is never used when we get /api/models request from UI. Right, and that happens only in two cases:
* first load of UI (this is your worry - we do not need to re-query endpoints if the key is fixed)
* after user key update (this is kind of point that we want to obtain)
Cache is used if some other part of code requests models using getModelsConfig()

We can say that all models retrieved with _not-user-provided_ key stay the same between different users. In this case we could to split endpoints using fixed keys and user defined. 

Then maybe we could cache fixed set and "user-provided". I guess it could not be simply between default models and custom models as both can use both fixed and user provided keys, right?

On the first glance the query of models are not easy to split up in this way. Probably doable, but it feels hardly worth.

Probably we could use caching of models per endpoint+key. Then we will need to do only 1 request - for the endpoint user just changed the key. Sounds the best for me. But this involves changes across all endpoints/providers.

Reply:
Looking a bit deeper, it looks like we cannot cache ALL models (as a block) like it is currently done in _ModelController_ if we need to allow different users to have a different set of models. 
User key that you can enter from UI implies exactly that. Say we have a custom configuration with a user provided key
* you get key with 5 models, you log in, enter the key, we cache N+5 models
* I get key with 2 models, I log in, enter the key, we cache N+2 models
* you do some operation, that requires model validation - if validation uses the cache, your operation will be validated against my 2 models :-(
The solution would be to cache models per url+key, and whenever we need a list of rebuild all the model set (and not use the fixed one in CONFIG_STORE). It looks like hardly any models get fetched from a remote endpoint except ones in custom config (and ones loaded via _getOpenAIModels_, that already caches models per endpoint)
So all the changes will go into _loadConfigModels_ (and I will remove any usage of cache in _ModelsController_) I will shortly update PR with these changes.
There is also some draft code in _overrideController()_ that intends to update the cache; this will need to aligned if/whenever implemented.